File: pkg/types/standings_test.go
package types

import (
	"testing"

	"github.com/google/go-cmp/cmp"

	"github.com/google/uuid"
)

func Test_MatchupsToStandingsMap(t *testing.T) {
	ms := Matchups{
		{
			ID:        uuid.NewString(),
			IsPlayoff: true,
		},
		{
			ID:         uuid.NewString(),
			HomeUserID: "1",
			AwayUserID: "2",
			HomeScore:  10,
			AwayScore:  5,
		},
		{
			ID:         uuid.NewString(),
			HomeUserID: "1",
			AwayUserID: "3",
			HomeScore:  20,
			AwayScore:  15,
		},
		{
			ID:         uuid.NewString(),
			HomeUserID: "2",
			AwayUserID: "3",
			HomeScore:  30,
			AwayScore:  25,
		},
		{
			ID:         uuid.NewString(),
			HomeUserID: "1",
			AwayUserID: "2",
			HomeScore:  15,
			AwayScore:  30,
		},
	}

	expected := StandingsMap{
		"1": {
			UserID:        "1",
			Wins:          2,
			Losses:        1,
			Ties:          0,
			PointsFor:     45,
			PointsAgainst: 50,
			H2HWins:       map[string]int{"2": 1, "3": 1},
		},
		"2": {
			UserID:        "2",
			Wins:          2,
			Losses:        1,
			Ties:          0,
			PointsFor:     65,
			PointsAgainst: 50,
			H2HWins:       map[string]int{"1": 1, "3": 1},
		},
		"3": {
			UserID:        "3",
			Wins:          0,
			Losses:        2,
			Ties:          0,
			PointsFor:     40,
			PointsAgainst: 50,
			H2HWins:       map[string]int{},
		},
	}

	if diff := cmp.Diff(expected, MatchupsToStandingsMap(ms)); diff != "" {
		t.Errorf("mismatch (-want +got):\n%s", diff)
	}
}

func Test_SortStandingsMap(t *testing.T) {
	standings := StandingsMap{
		"1": {
			UserID:        "1",
			Wins:          8,
			Losses:        4,
			Ties:          0,
			PointsFor:     1808.65,
			PointsAgainst: 1544.05,
			H2HWins:       map[string]int{},
		},
		"2": {
			UserID:        "2",
			Wins:          2,
			Losses:        10,
			Ties:          0,
			PointsFor:     1552.05,
			PointsAgainst: 1731.15,
			H2HWins:       map[string]int{},
		},
		"3": {
			UserID:        "3",
			Wins:          6,
			Losses:        6,
			Ties:          0,
			PointsFor:     1495.05,
			PointsAgainst: 1596.55,
			H2HWins: map[string]int{
				"5": 1,
			},
		},
		"4": {
			UserID:        "4",
			Wins:          6,
			Losses:        6,
			Ties:          0,
			PointsFor:     1584.10,
			PointsAgainst: 1676.55,
			H2HWins: map[string]int{
				"3": 1,
			},
		},
		"5": {
			UserID:        "5",
			Wins:          6,
			Losses:        6,
			Ties:          0,
			PointsFor:     1584.10,
			PointsAgainst: 1676.55,
			H2HWins:       map[string]int{},
		},
	}

	expected := Standings{
		standings["1"],
		standings["4"],
		standings["3"],
		standings["5"],
		standings["2"],
	}

	if diff := cmp.Diff(expected, standings.SortStandingsMap()); diff != "" {
		t.Errorf("mismatch (-want +got):\n%s", diff)
	}
}

File: worker/src/domain/standings.test.ts
import { describe, expect, it } from "vitest";
import { matchupsToStandingsMap, sortStandingsMap, standingsForLeague } from "./standings";
import { League, LeagueStatus, Matchup, PlayoffRound } from "./types";

let nextId = 0;

function matchup(overrides: Partial<Matchup> & Pick<Matchup, "home_user_id" | "away_user_id">): Matchup {
  return {
    id: `m${nextId++}`,
    year: 2025,
    week: 1,
    is_playoff: false,
    playoff_round: null,
    home_seed: null,
    away_seed: null,
    home_score: 100,
    away_score: 90,
    ...overrides,
  };
}

function league(status: string): League {
  return {
    id: "L1",
    year: 2025,
    first_place: "",
    second_place: "",
    third_place: "",
    status,
  };
}

describe("matchupsToStandingsMap", () => {
  it("accumulates records, points, and head-to-head wins", () => {
    const standings = matchupsToStandingsMap([
      matchup({ home_user_id: "a", away_user_id: "b", home_score: 100, away_score: 90 }),
      matchup({ home_user_id: "b", away_user_id: "a", week: 2, home_score: 120, away_score: 80 }),
      matchup({ home_user_id: "a", away_user_id: "b", week: 3, home_score: 50, away_score: 50 }),
    ]);

    const a = standings.get("a")!;
    expect(a.wins).toBe(1);
    expect(a.losses).toBe(1);
    expect(a.ties).toBe(1);
    expect(a.pointsFor).toBe(100 + 80 + 50);
    expect(a.pointsAgainst).toBe(90 + 120 + 50);
    expect(a.h2hWins.get("b")).toBe(1);
    expect(standings.get("b")!.h2hWins.get("a")).toBe(1);
  });

  it("ignores playoff matchups", () => {
    const standings = matchupsToStandingsMap([
      matchup({ home_user_id: "a", away_user_id: "b" }),
      matchup({
        home_user_id: "a",
        away_user_id: "b",
        week: 15,
        is_playoff: true,
        playoff_round: PlayoffRound.Finals,
      }),
    ]);
    expect(standings.get("a")!.wins).toBe(1);
  });
});

describe("sortStandingsMap tiebreakers", () => {
  it("orders by wins first", () => {
    const standings = matchupsToStandingsMap([
      matchup({ home_user_id: "winner", away_user_id: "loser" }),
      matchup({ home_user_id: "winner", away_user_id: "loser", week: 2 }),
    ]);
    expect(sortStandingsMap(standings).map((s) => s.userId)).toEqual(["winner", "loser"]);
  });

  it("breaks a tie on head-to-head before points for", () => {
    // "a" and "b" both finish 1-1, and they are the only teams on one win.
    // "b" beat "a" head-to-head while scoring far fewer points overall, so
    // head-to-head has to outrank points for.
    const standings = matchupsToStandingsMap([
      matchup({ home_user_id: "b", away_user_id: "a", home_score: 80, away_score: 70 }),
      matchup({ home_user_id: "a", away_user_id: "d", week: 2, home_score: 200, away_score: 10 }),
      matchup({ home_user_id: "e", away_user_id: "b", week: 3, home_score: 100, away_score: 10 }),
      matchup({ home_user_id: "e", away_user_id: "d", week: 4, home_score: 100, away_score: 10 }),
    ]);

    expect(standings.get("a")!.pointsFor).toBeGreaterThan(standings.get("b")!.pointsFor);

    const oneWin = sortStandingsMap(standings).filter((s) => s.wins === 1);
    expect(oneWin.map((s) => s.userId)).toEqual(["b", "a"]);
  });

  it("falls back to points for when head-to-head is level", () => {
    const standings = matchupsToStandingsMap([
      matchup({ home_user_id: "high", away_user_id: "x", home_score: 150, away_score: 10 }),
      matchup({ home_user_id: "low", away_user_id: "y", home_score: 100, away_score: 10 }),
    ]);
    const sorted = sortStandingsMap(standings).filter((s) => s.wins === 1);
    expect(sorted.map((s) => s.userId)).toEqual(["high", "low"]);
  });
});

describe("standingsForLeague", () => {
  const regularSeason: Matchup[] = [];
  const teams = ["t1", "t2", "t3", "t4", "t5", "t6", "t7", "t8"];
  // Give every team a distinct record: t1 best, t8 worst.
  teams.forEach((team, i) => {
    for (let w = 0; w < teams.length - i; w++) {
      regularSeason.push(
        matchup({
          home_user_id: team,
          away_user_id: "bye",
          week: w + 1,
          home_score: 100,
          away_score: 50,
        }),
      );
    }
  });

  it("throws for a pending league", () => {
    expect(() => standingsForLeague(league(LeagueStatus.Pending), [])).toThrow(
      /has not started/,
    );
  });

  it("uses regular season order for an in-progress league", () => {
    const { standings, usedPlayoffResults } = standingsForLeague(
      league(LeagueStatus.InProgress),
      regularSeason,
    );
    expect(usedPlayoffResults).toBe(false);
    expect(standings[0]!.userId).toBe("t1");
  });

  it("uses playoff results for the top six of a complete league", () => {
    const playoffs: Matchup[] = [
      matchup({
        home_user_id: "t4",
        away_user_id: "t5",
        week: 15,
        is_playoff: true,
        playoff_round: PlayoffRound.Quarterfinals,
        home_score: 100,
        away_score: 90,
      }),
      matchup({
        home_user_id: "t3",
        away_user_id: "t6",
        week: 15,
        is_playoff: true,
        playoff_round: PlayoffRound.Quarterfinals,
        home_score: 100,
        away_score: 90,
      }),
      matchup({
        home_user_id: "t2",
        away_user_id: "t1",
        week: 17,
        is_playoff: true,
        playoff_round: PlayoffRound.Finals,
        home_score: 100,
        away_score: 90,
      }),
      matchup({
        home_user_id: "t3",
        away_user_id: "t4",
        week: 17,
        is_playoff: true,
        playoff_round: PlayoffRound.ThirdPlace,
        home_score: 100,
        away_score: 90,
      }),
    ];

    const { standings, usedPlayoffResults } = standingsForLeague(
      league(LeagueStatus.Complete),
      [...regularSeason, ...playoffs],
    );

    expect(usedPlayoffResults).toBe(true);
    // Finals winner, finals loser, third place winner, third place loser,
    // then the two quarterfinal losers.
    expect(standings.slice(0, 4).map((s) => s.userId)).toEqual(["t2", "t1", "t3", "t4"]);
    expect(standings.slice(4, 6).map((s) => s.userId).sort()).toEqual(["t5", "t6"]);
  });

  it("falls back to regular season order when a complete league has no playoff data", () => {
    const { standings, usedPlayoffResults } = standingsForLeague(
      league(LeagueStatus.Complete),
      regularSeason,
    );
    expect(usedPlayoffResults).toBe(false);
    expect(standings[0]!.userId).toBe("t1");
  });
});
